### PR TITLE
constrain TokenType JAR

### PR DESCRIPTION
**Pull request by roastario**

### CODE REVIEW

**Comment by tudor-malene** [comment]
I was expecting a new field added to the `AbstractToken` state. Which would have been propagated from input to output states, and then the jar from the transaction attachments checked against it.

Unless I'm missing something, the code in this pr just reads the jar from the classloader, and then checks it against all the other states that read from the same classloader, which means there's no security check on it.

**Comment by roastario** [comment]
> I was expecting a new field added to the `AbstractToken` state. Which would have been propagated from input to output states, and then the jar from the transaction attachments checked against it.
> 
> Unless I'm missing something, the code in this pr just reads the jar from the classloader, and then checks it against all the other states that read from the same classloader, which means there's no security check on it.

@tudor-malene We do enforce the check that the jar hash never changes via the `com.r3.corda.lib.tokens.contracts.AbstractTokenContract#verifyAllTokensUseSameTypeJar`
which enforces the fact that all inputs and outputs *must* use the same hash (and jar)

**Comment by tudor-malene** [comment]
> > I was expecting a new field added to the `AbstractToken` state. Which would have been propagated from input to output states, and then the jar from the transaction attachments checked against it.
> > Unless I'm missing something, the code in this pr just reads the jar from the classloader, and then checks it against all the other states that read from the same classloader, which means there's no security check on it.
> 
> @tudor-malene We do enforce the check that the jar hash never changes via the `com.r3.corda.lib.tokens.contracts.AbstractTokenContract#verifyAllTokensUseSameTypeJar`
> which enforces the fact that all inputs and outputs _must_ use the same hash (and jar)

But of course it doesn't change if you read the jar from the same classloader for both the input and the output states.

**Comment by roastario** [comment]
> > > I was expecting a new field added to the `AbstractToken` state. Which would have been propagated from input to output states, and then the jar from the transaction attachments checked against it.
> > > Unless I'm missing something, the code in this pr just reads the jar from the classloader, and then checks it against all the other states that read from the same classloader, which means there's no security check on it.
> > 
> > 
> > @tudor-malene We do enforce the check that the jar hash never changes via the `com.r3.corda.lib.tokens.contracts.AbstractTokenContract#verifyAllTokensUseSameTypeJar`
> > which enforces the fact that all inputs and outputs _must_ use the same hash (and jar)
> 
> But of course it doesn't change if you read the jar from the same classloader for both the input and the output states.


OK, maybe I am missing something but 

Issuer issues token with hash = [ABCFD]

from then on all tokens of this type must be also [ABCFD] no?

The detail is that the tokens contract will "group" all tokens of a given type before applying this rule - so it should be possible to enforce that the token jar must always be the same as when it was issued.

**Comment by tudor-malene** [comment]
If you don’t encode the hash in the state, you have to load the previous transaction of all input states and read the hash from that classloader

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by roastario** [comment]
> If you don’t encode the hash in the state, you have to load the previous transaction of all input states and read the hash from that classloader

Ah, I realise what you mean now. 

Yes sorry, I have changed it to a `val` instead of a `fun` in the current version of the code, I forgot to repush.

File: contracts/src/main/kotlin/com/r3/corda/lib/tokens/contracts/states/FungibleToken.kt [hunk 1]
```diff
@@ -31,9 +33,10 @@ import net.corda.core.schemas.QueryableState
  * @param T the [TokenType] this [FungibleToken] state is in respect of.
  */
 @BelongsToContract(FungibleTokenContract::class)
-open class FungibleToken<T : TokenType>(
+class FungibleToken<T : TokenType>(
```
Comment (kasiastreich) [inline contracts/src/main/kotlin/com/r3/corda/lib/tokens/contracts/states/FungibleToken.kt]: NonFungibleToken is still open with one final field is it on purpose?

File: deterministic.gradle [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
-private final void deterministic(Configuration configuration) {
-    if (configuration.state == Configuration.State.UNRESOLVED) {
-        // Ensure that this module uses the deterministic Corda artifacts.
-        configuration.resolutionStrategy.dependencySubstitution {
-            substitute module("$corda_release_group:corda-serialization:$corda_release_version") with module("$corda_release_group:corda-serialization-deterministic:$corda_release_version")
-            substitute module("$corda_release_group:corda-core:$corda_release_version") with module("$corda_release_group:corda-core-deterministic:$corda_release_version")
-        }
-    }
-}
+//configurations {
```
Comment (kasiastreich) [inline deterministic.gradle]: cleanup

File: modules/test-contracts/src/main/kotlin/com/r3/corda/lib/tokens/testing/states/Rubles.kt [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package com.r3.corda.lib.tokens.testing.states
+
+import com.r3.corda.lib.tokens.contracts.types.TokenType
+
+
+open class Ruble : TokenType {
```
Comment (kasiastreich) [inline modules/test-contracts/src/main/kotlin/com/r3/corda/lib/tokens/testing/states/Rubles.kt]: oh man, wrong currency, polski złoty

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/issue/IssueTokensUtilities.kt [hunk 4]
```diff
@@ -45,4 +46,28 @@ fun addIssueTokens(transactionBuilder: TransactionBuilder, vararg outputs: Abstr
 @Suspendable
 fun addIssueTokens(transactionBuilder: TransactionBuilder, output: AbstractToken<*>): TransactionBuilder {
     return addIssueTokens(transactionBuilder, listOf(output))
+}
+
+// Utilities for ensuring that the correct JAR which implements TokenType is added to the transaction.
+
+fun addTokenTypeJar(tokens: List<AbstractToken<*>>, transactionBuilder: TransactionBuilder) {
```
Comment (kasiastreich) [inline workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/issue/IssueTokensUtilities.kt]: Should we move it to internal package?

File: workflows/src/test/kotlin/com/r3/corda/lib/tokens/workflows/JITMockNetworkTests.kt [hunk 5]
```diff
@@ -85,4 +88,11 @@ abstract class JITMockNetworkTests(val names: List<CordaX500Name> = emptyList())
      * Smart helper for [assertNotHasTransaction] that passes in the test suite's network.
      */
     protected fun assertNotHasTransaction(tx: SignedTransaction, vararg nodes: StartedMockNode) = assertNotHasTransaction(tx, network, *nodes)
+
+}
+
+fun convertToStartedMockNode(createPartyNode: TestStartedNode): StartedMockNode {
```
Comment (kasiastreich) [inline workflows/src/test/kotlin/com/r3/corda/lib/tokens/workflows/JITMockNetworkTests.kt]: What is it?

File: contracts/src/main/kotlin/com/r3/corda/lib/tokens/contracts/states/FungibleToken.kt (hunk 1 above)
Comment (roger3cev) [reply contracts/src/main/kotlin/com/r3/corda/lib/tokens/contracts/states/FungibleToken.kt]: We can open it up later if necessary.

File: modules/test-contracts/src/main/kotlin/com/r3/corda/lib/tokens/testing/states/Rubles.kt (hunk 3 above)
Comment (roger3cev) [reply modules/test-contracts/src/main/kotlin/com/r3/corda/lib/tokens/testing/states/Rubles.kt]: :laugh

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/issue/IssueTokensUtilities.kt (hunk 4 above)
Comment (roger3cev) [reply workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/issue/IssueTokensUtilities.kt]: Others that create their own flows might need to use this

File: workflows/src/test/kotlin/com/r3/corda/lib/tokens/workflows/JITMockNetworkTests.kt (hunk 5 above)
Comment (roger3cev) [reply workflows/src/test/kotlin/com/r3/corda/lib/tokens/workflows/JITMockNetworkTests.kt]: @roastario  ?

**Comment by roger3cev** [comment]
Looks like tests are passing and we have approval from @tudor-malene !

**Pull request merged**